[SMALLFIX] Fix a race condition when concurrently creating blocks on …

apc999 opened the pull request:
…worker

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/10559/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/10560/
Test FAILed.

gpang commented:
Jenkins, test this please

aaudiber commented on core/common/src/main/java/alluxio/util/io/FileUtils.java:
```diff
@@ -218,15 +218,19 @@ public FileVisitResult postVisitDirectory(Path dir, IOException e) throws IOExce
   public static void createStorageDirPath(String path) throws IOException {
     File dir = new File(path);
     String absolutePath = dir.getAbsolutePath();
```
we can move this declaration closer to its usage

apc999 replied:
done

aaudiber commented on core/common/src/test/java/alluxio/util/io/FileUtilsTest.java:
```diff
@@ -256,4 +260,57 @@ public void createDirTest() throws IOException {
+    File storageDir = new File(mTestFolder.getRoot(), "tmp");
+    File blockFile = new File(storageDir, "200");
+
+    // When storage dir not exists
```
not exists -> doesn't exist

apc999 replied:
done

aaudiber commented on core/common/src/test/java/alluxio/util/io/FileUtilsTest.java:
```diff
@@ -256,4 +260,57 @@ public void createDirTest() throws IOException {
   }
+
+  /**
+   * Tests {@link FileUtils#createBlockPath} method when storage dir exists or not.
```
not -> doesn't exist

apc999 replied:
done

aaudiber commented on core/common/src/test/java/alluxio/util/io/FileUtilsTest.java:
```diff
@@ -256,4 +260,57 @@ public void createDirTest() throws IOException {
+   */
+  @Test
+  public void createStorageDirPathTest() throws IOException {
+    File storageDir = new File(mTestFolder.getRoot(), "tmp");
```
tmp -> storageDir

apc999 replied:
done

aaudiber commented on core/common/src/test/java/alluxio/util/io/FileUtilsTest.java:
```diff
@@ -256,4 +260,57 @@ public void createDirTest() throws IOException {
+      }
+    }
+
+    for (int i = 0; i < 100; i++) {
```
How long does this take to run? Does it usually fail without your change?

apc999 replied:
This test is very fast. But without the change in `FileUtils`, I cannot reproduce the bug.

aaudiber replied:
If we want to really stress concurrency, it might help to run 10 threads at the same time and use a cyclic barrier or countdown latch to coordinate their calls to createBlockPath

apc999 replied:
Updated. Now it is very easy to reproduce the original bug

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/10562/
Test PASSed.

aaudiber commented:
Thanks for fixing this critical race condition @apc999! I left some comments on the code

apc999 commented:
@aaudiber I addressed your comments. One concern might be that the concurrent test is tricky---I could not reproduce the original bug either through this test.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/10563/
Test PASSed.

aaudiber commented:
@apc999 I left a suggestion for how to make the concurrency test stronger. Feel free to address it or punt. Otherwise LGTM

aaudiber commented on core/common/src/test/java/alluxio/util/io/FileUtilsTest.java:
```diff
@@ -262,14 +262,14 @@ public void createDirTest() throws IOException {
     File blockFile = new File(storageDir, "200");
 
-    // When storage dir not exists
+    // When storage dir doesn't exists
```
exists -> exist

apc999 replied:
done

apc999 commented:
@aaudiber  PTAL

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/10567/
Test PASSed.

aaudiber commented on core/common/src/test/java/alluxio/util/io/FileUtilsTest.java:
```diff
@@ -256,4 +263,67 @@ public void createDirTest() throws IOException {
+   * when multiple blocks belonging to the same storage dir get created concurrently.
+   */
+  @Test
+  public void concurrentCreateStorageDirPathTest() throws Exception {
```
Does it fail now without the change? It would only take 2 lines to add CyclicBarrier to improve concurrency,

```java
final CyclicBarrier barrier = new CyclicBarrier(numCreators);
```
at the start, and
```java
barrier.await();
```
right before calling createBlockPath.

https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/CyclicBarrier.html

apc999 replied:
Updated

aaudiber commented:
LGTM

apc999 commented:
@aaudiber thanks for the suggestion. I have improved the tests using `CyclicBarrier`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/10569/
Test PASSed.
